### PR TITLE
Triggering provazio controller instead of manually doing stuff

### PR DESCRIPTION
**Pull request by Hedingber**
Moving the logic on how scale to/from zero to provazio, adding logic to trigger provazio to do the job + supporting new scaler types structure (https://github.com/v3io/scaler-types/pull/2)

### CODE REVIEW

File: resourcescaler.go [hunk 1]
```diff
@@ -48,147 +47,142 @@ func New(kubeconfigPath string, namespace string) (scaler_types.ResourceScaler,
...
 }
 
-func (s *AppResourceScaler) SetScale(resource scaler_types.Resource, scaling int) error {
-
-	// get ingress by resource name
-	ingress, err := s.kubeClientSet.ExtensionsV1beta1().Ingresses(s.namespace).Get(string(resource), meta_v1.GetOptions{})
-	if err != nil {
-		s.logger.WarnWith("Failure during retrieval of ingress", "resource_name", string(resource))
-		return errors.Wrap(err, "Failed getting ingress instance")
+func (s *AppResourceScaler) SetScale(resource scaler_types.Resource, scale int) error {
```
Comment (omesser) [inline resourcescaler.go]: Looks like scale should be a bool or enum with 2 values, wince this doesn't really support anything else beside to/from zero *by design*

File: resourcescaler.go [hunk 2]
```diff
@@ -48,147 +47,142 @@ func New(kubeconfigPath string, namespace string) (scaler_types.ResourceScaler,
...
 
-	ingress.GetObjectMeta().SetAnnotations(map[string]string{
-		"nginx.ingress.kubernetes.io/configuration-snippet": fmt.Sprintf(
-			`proxy_set_header X-App-Target "%s";`, string(resource)),
+func (s *AppResourceScaler) scaleServiceFromZero(namespace string, serviceName string) error {
+	var jsonPatchMapper []map[string]string
+	s.logger.DebugWith("Scaling from zero", "namespace", namespace, "serviceName", serviceName)
+	path := fmt.Sprintf("/spec/spec/tenants/0/spec/services/%s/state", string(serviceName))
+	jsonPatchMapper = append(jsonPatchMapper, map[string]string{
+		"op":    "add",
```
Comment (omesser) [inline resourcescaler.go]: update?

File: resourcescaler.go [hunk 3]
```diff
@@ -48,147 +47,142 @@ func New(kubeconfigPath string, namespace string) (scaler_types.ResourceScaler,
...
-		}
-	} else {
-		// update service selector to refer to resource
-		s.logger.InfoWith("Changing service's selector back to work with resource", "service_name", string(resource))
+func (s *AppResourceScaler) scaleServiceToZero(namespace string, serviceName string) error {
+	var jsonPatchMapper []map[string]string
+	s.logger.DebugWith("Scaling to zero", "namespace", namespace, "serviceName", serviceName)
+	path := fmt.Sprintf("/spec/spec/tenants/0/spec/services/%s/state", string(serviceName))
+	jsonPatchMapper = append(jsonPatchMapper, map[string]string{
+		"op":    "add",
```
Comment (omesser) [inline resourcescaler.go]: update?

File: resourcescaler.go [hunk 4]
```diff
@@ -48,147 +47,142 @@ func New(kubeconfigPath string, namespace string) (scaler_types.ResourceScaler,
...
 	}
+	return nil
+}
 
-	// if scaling up, make sure that all pods are in running state
-	if scaling != 0 {
-		s.logger.Info("Waiting for pods to be running")
-		if s.waitForServicePodsStatus(service, s.namespace, v1.PodRunning) != nil {
-			return errors.Wrap(err, "Failed while waiting for service pods status")
+func (s *AppResourceScaler) waitForServiceReadiness(namespace string, serviceName string) error {
```
Comment (omesser) [inline resourcescaler.go]: Why is this needed?

File: resourcescaler.go [hunk 5]
```diff
@@ -48,147 +47,142 @@ func New(kubeconfigPath string, namespace string) (scaler_types.ResourceScaler,
...
 		}
-		s.logger.Info("All pods are running")
+		time.Sleep(5 * time.Second)
 	}
-	
 	return nil
 }
 
 func (s *AppResourceScaler) GetResources() ([]scaler_types.Resource, error) {
+	var iguazioTenantAppServicesSetMap map[string]interface{}
```
Comment (omesser) [inline resourcescaler.go]: This function doesn't `GetResources()` but `GetIguazioTenantAppAerviceAets()`
Also comment above about why needed

File: resourcescaler.go (hunk 1 above)
Comment (Hedingber) [reply resourcescaler.go]: Mmm yup, but changing it requires `nuclio` `scaler` and `app-resource-scaler` PR so lets postpone it for now

File: resourcescaler.go (hunk 2 above)
Comment (Hedingber) [reply resourcescaler.go]: Talked offline, there's no update op(eration), there's maybe `replace` but it will fail if the field doesn't exist, `add` will create the field if it doesn't exist, or replace its value if it exists

File: resourcescaler.go (hunk 3 above)
Comment (Hedingber) [reply resourcescaler.go]: same

File: resourcescaler.go (hunk 4 above)
Comment (Hedingber) [reply resourcescaler.go]: Talked offline - when setScale return in the state of scaling from zero the DLX will release the queue of messages, therefore we need to return only when the service is really ready

File: resourcescaler.go (hunk 5 above)
Comment (Hedingber) [reply resourcescaler.go]: Mmm `GetResources` is the name of the method in the resourceScaler interface that I'm implementing for app services, in which I practically get `IguazioTenantAppServiceSets`
https://github.com/v3io/scaler-types/blob/master/types.go#L37

File: resourcescaler.go [hunk 6]
```diff
@@ -113,82 +235,244 @@ func (s *AppResourceScaler) patchIguazioTenantAppServiceSets(namespace string, j
 	return nil
 }
 
-func (s *AppResourceScaler) waitForServiceReadiness(namespace string, serviceName string) error {
-	s.logger.DebugWith("Waiting for service readiness", "serviceName", serviceName)
+func (s *AppResourceScaler) waitForServiceState(namespace string, serviceName string, state string) error {
+	s.logger.DebugWith("Waiting for service to reach state", "serviceName", serviceName, "state", state)
 	for {
```
Comment (omesser) [inline resourcescaler.go]: This can be stuck forever if something goes wrong (like a service is restarted or goes into error state mid scaling to/from zero.
Suggest some reasonable long timeout and using a `select` with a ticker and a deadline

File: resourcescaler.go [hunk 7]
```diff
@@ -113,82 +235,244 @@ func (s *AppResourceScaler) patchIguazioTenantAppServiceSets(namespace string, j
...
+		return []scaler_types.ScaleResource{}, nil
+	}
+
+	scaleToZeroMode, ok := scaleToZeroSpec["mode"].(string)
+	if !ok {
+		return []scaler_types.ScaleResource{}, errors.New("Scale to zero spec does not have mode")
+	}
+
+	// if it's not enabled there's no reason to parse the rest
+	if scaleToZeroMode == "enabled" {
```
Comment (omesser) [inline resourcescaler.go]: Can be nice to check if `!= 'enabled'` and return early, then the rest of this won't be indented
